kotlinx-coroutines-core/common/src/CoroutineStart.kt

```diff
@@ -1,74 +1,343 @@
+     * Example:
+     * ```
+     * val N_PERMITS = 3
+     * val semaphore = Semaphore(N_PERMITS)
+     * try {
+     *     repeat(100) {
+     *         semaphore.acquire()
+     *         if (it != 7) {
+     *             println("Scheduling $it...")
+     *         } else {
+     *             // "randomly" cancel the whole procedure
+     *             cancel()
+     *         }
+     *         launch(Dispatchers.Default, start = CoroutineStart.ATOMIC) {
+     *             println("Entered $it")
+     *             try {
+     *                 // this `try` block will be entered in any case because of ATOMIC
+     *                 println("Performing the procedure $it")
+     *                 delay(10.milliseconds) // may throw due to cancellation
+     *                 println("Done with the procedure $it")
+     *             } finally {
+     *                 semaphore.release()
+     *             }
+     *         }
+     *     }
+     * } finally {
+     *     withContext(NonCancellable) {
+     *         repeat(N_PERMITS) { semaphore.acquire() }
+     *         println("All permits were successfully returned!")
+     *     }
+     * }
+     * ```
+     *
+     * Here, we used [ATOMIC] to ensure that a semaphore that was acquired outside of the coroutine does get released
+     * even if cancellation happens between `acquire()` and `launch`.
+     * As a result, the semaphore will eventually regain all three permits.
+     *
+     * Behavior of [ATOMIC] can be described with the following examples:
+     *
+     * ```
+     * // Example of cancelling atomically started coroutines
+     * runBlocking {
+     *     println("1. Atomically starting a coroutine that goes through a dispatch.")
+     *     launch(start = CoroutineStart.ATOMIC) {
+     *         check(!isActive) // attempting to suspend will throw
+     *         println("4. A coroutine that went through a dispatch also starts.")
+     *         try {
+     *             delay(10.milliseconds)
+     *             println("This code will never run.")
+     *         } catch (e: CancellationException) {
+     *             println("5. Cancellation at later points still works.")
+     *             throw e
+     *         }
+     *     }
+     *     println("2. Cancelling this coroutine and all of its children.")
+     *     cancel()
+     *     launch(Dispatchers.Unconfined, start = CoroutineStart.ATOMIC) {
+     *         check(!isActive) // attempting to suspend will throw
+     *         println("3. An undispatched coroutine starts.")
+     *     }
+     *     ensureActive() // we can even crash the current coroutine.
+     * }
+     * ```
+     *
+     * **Pitfall**: occasionally, it is an error to execute code after the coroutine was cancelled.
+     * A notable example is working with UI on Android:
+     * after a coroutine scope whose lifecycle is tied to to an UI element gets cancelled,
+     * updating that UI element is not allowed and will throw an exception.
```

Comment:
Do we have a source for this / would it make sense to link the corresponding discussion here? (https://github.com/Kotlin/kotlinx.coroutines/issues/3504#issuecomment-1294865970)

Reply:
I actually don't think there is. That discussion highlights the general problem, but ultimately, it's unrelated to `CoroutineStart`.